File: strings/medium/java_solutions/WordSearch_79.java
package strings.medium.java_solutions;

import java.util.Arrays;

public class WordSearch_79 {

    private static final int[][] DIR = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    private static final char REPLACE_SYMBOL = '~';

    public boolean exist(char[][] board, String word) {
        if (word.length() == 0) {
            return true;
        }
        for (var i = 0; i < board.length; i++) {
            for (var j = 0; j < board[0].length; j++) {
                if (explore(board, word, i, j)) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean explore(char[][] board, String word, int i, int j) {
        if (word.length() == 1) {
            return word.charAt(0) == board[i][j];
        }
        if (word.charAt(0) != board[i][j]) {
            return false;
        }
        board[i][j] = REPLACE_SYMBOL;
        for (int[] direction : DIR) {
            var nextRow = i + direction[0];
            var nextColumn = j + direction[1];

            if (nextRow < 0 || nextColumn < 0 || nextRow >= board.length || nextColumn >= board[0].length || board[nextRow][nextColumn] == REPLACE_SYMBOL) {
                continue;
            }
            if (explore(board, word.substring(1), nextRow, nextColumn)) {
                return true;
            }
        }
        board[i][j] = word.charAt(0);
        return false;
    }
}

File: leetcode/strings/medium/golang_solutions/word_search_79.go
package golang_solutions

var directions = [][]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
var replaceSymbol byte = 126

func exist(board [][]byte, word string) bool {
	if len(word) == 0 {
		return true
	}
	for i := range board {
		for j := range board[i] {
			if explore(board, word, i, j) {
				return true
			}
		}
	}
	return false
}

func explore(board [][]byte, word string, i, j int) bool {
	if len(word) == 1 {
		return word[0] == board[i][j]
	}
	if board[i][j] != word[0] {
		return false
	}
	board[i][j] = replaceSymbol
	for _, dir := range directions {
		nextRow := i + dir[0]
		nextColumn := j + dir[1]

		if nextRow < 0 || nextColumn < 0 || nextRow >= len(board) || nextColumn >= len(board[0]) || board[nextRow][nextColumn] == replaceSymbol {
			continue
		}
		if explore(board, word[1:], nextRow, nextColumn) {
			return true
		}
	}
	board[i][j] = word[0]
	return false
}
